pkg/chain/ethereum/ethutil/rate_limiter.go

```diff
@@ -0,0 +1,218 @@
+package ethutil
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"golang.org/x/sync/semaphore"
+	"golang.org/x/time/rate"
+	"math/big"
+	"time"
+)
+
+type rateLimiter struct {
+	bind.ContractBackend
+
+	limiter   *rate.Limiter
+	semaphore *semaphore.Weighted
```

Comment:
I am not sure if we should have a `semaphore` and `ConcurrentLimit` in the rate limiter given [there is no concurrency in ethereum implementation](https://github.com/keep-network/keep-core/blob/master/pkg/chain/ethereum/connect.go#L46-L58):
```
// transactionMutex allows interested parties to forcibly serialize
// transaction submission.
//
// When transactions are submitted, they require a valid nonce. The nonce is
// equal to the count of transactions the account has submitted so far, and
// for a transaction to be accepted it should be monotonically greater than
// any previous submitted transaction. To do this, transaction submission
// asks the Ethereum client it is connected to for the next pending nonce,
// and uses that value for the transaction. Unfortunately, if multiple
// transactions are submitted in short order, they may all get the same
// nonce. Serializing submission ensures that each nonce is requested after
// a previous transaction has been submitted.
transactionMutex *sync.Mutex
```

Even if users set `ConcurrencyLimit` to `>1` in their configuration, they simply can't get it.

I'd leave the concurrency part to the underlying implementation and focus only on rate-limiting here.

Reply:
Yeah, transactions have no concurrency but read-only calls have. And the concurrency of calls was the main problem we observed on Alchemy.

Reply:
Ah, it makes perfect sense. I think we should improve the documentation as it wasn't clear to me at all.